vllm/worker/model_runner.py

```diff
@@ -645,82 +644,95 @@ def _prepare_model_input(
+            "selected_token_indices": sampling_metadata.selected_token_indices,
+            "lora_requests": lora_requests,
+            "lora_mapping": lora_mapping,
+            "multi_modal_kwargs": multi_modal_kwargs,
+            "num_prefill_tokens": num_prefill_tokens,
+            "num_decode_tokens": num_decode_tokens,
+            "slot_mapping": slot_mapping,
+            "num_prefills": num_prefills,
+        }
+        if attn_metadata:
+            metadata_dict.update(attn_metadata.asdict_zerocopy())
+
+        # sampling_metadata should not appear in metadata_dict.
+        # but we need to access it later.
+        self.sampling_metadata = sampling_metadata
+
+        return metadata_dict
+
+    def convert_tensor_dict_to_model_input(self, metadata_dict: Dict[str,
+                                                                     Any]):
+        input_tokens = metadata_dict.pop("input_tokens")
+        input_positions = metadata_dict.pop("input_positions")
+        selected_token_indices = metadata_dict.pop("selected_token_indices")
+        lora_mapping = metadata_dict.pop("lora_mapping")
+        lora_requests = metadata_dict.pop("lora_requests")
+        multi_modal_kwargs = metadata_dict.pop("multi_modal_kwargs")
+        if metadata_dict:
+            attn_metadata = self.attn_backend.make_metadata(**metadata_dict)
+        else:
+            attn_metadata = None
+
+        if hasattr(self, "sampling_metadata"):
+            sampling_metadata = self.sampling_metadata
         else:
-            metadata_dict = broadcast_tensor_dict(src=0)
-            input_tokens = metadata_dict.pop("input_tokens")
-            input_positions = metadata_dict.pop("input_positions")
-            selected_token_indices = metadata_dict.pop(
-                "selected_token_indices")
-            lora_mapping = metadata_dict.pop("lora_mapping")
-            lora_requests = metadata_dict.pop("lora_requests")
-            multi_modal_kwargs = metadata_dict.pop("multi_modal_kwargs")
-            if metadata_dict:
-                attn_metadata = self.attn_backend.make_metadata(
-                    **metadata_dict)
-            else:
-                attn_metadata = None
             sampling_metadata = SamplingMetadata(
                 seq_groups=None,
                 selected_token_indices=selected_token_indices,
                 categorized_sample_indices=None,
                 num_prompts=0,
             )
-
         return (input_tokens, input_positions, attn_metadata,
                 sampling_metadata, lora_requests, lora_mapping,
                 multi_modal_kwargs)
 
+    def prepare_input_tensors(
+        self,
+        seq_group_metadata_list: Optional[List[SequenceGroupMetadata]],
+    ) -> Tuple[torch.Tensor, torch.Tensor, AttentionMetadata, SamplingMetadata,
+               Set[LoRARequest], LoRAMapping, Dict[str, torch.Tensor]]:
+        # TODO: deprecate this function. It is only used in tests.
```

Comment:
Can we directly deprecate this function in this PR?

Reply:
yes we can, if you like.

Reply:
let's do it!

Reply:
not yet. will do it after we finalize the design. I'm still waiting for the overall review on this pr.

Reply:
this function is kept for `profile_run`, no need to deprecate it now.